### PR TITLE
Preserve Python dispatch keys upon copy_tensor_metadata_except_version_counter

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #75644

Whether or not this is a reasonable operation to do in the presence of
subclasses is a good question in and of itself, but this fixes an
obvious invariant violation, which is that if a Tensor reports that
it is a tensor subclass, it had better have the Python dispatch key.
Previously, the dispatch key would have gotten unconditionally cleared;
now we preserve what ever the original bit was.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

### CODE REVIEW

File: test/test_python_dispatch.py [hunk 1]
```diff
@@ -706,6 +706,27 @@ def __torch_dispatch__(cls, func, types, args=(), kwargs=None):
...
+            @classmethod
+            def __torch_dispatch__(cls, func, types, args=(), kwargs=None):
+                nonlocal called
+                called += 1
+                return super().__torch_dispatch__(func, types, args, kwargs)
+
+        x = SubTensor(torch.empty(2))
+        x.data
+        self.assertEqual(called, 1)
+        x.data = torch.empty(2)
```
Comment (albanD) [inline test/test_python_dispatch.py:724]: I would expect set_ to have the same behavior here (and also be fixed by this PR) could you add a test for that?

**Review by albanD — COMMENTED** [review]
I would highly question any use of .data in general (and in the presence of subclass even more!)
But set_ should work and I think this is fixing it?

File: test/test_python_dispatch.py (hunk 1 above)
Comment (ezyang) [reply test/test_python_dispatch.py:724]: I think set_ already worked, because it gets handled by torch dispatch already

**Review by albanD — APPROVED** [review]
Thanks!

**Pull request closed without merging**